dobutsu.6
.TH DOBUTSU 6 "2021-01-16" "Robert Clausecker" "Games"
.
.SH NAME
\fBdobutsu\fR \(en play Dobutsu Shogi
.
.SH SYNOPSIS
\fBdobutsu\fR
[-\fBqv\fR]
[-\fBc \fIcolor\fR]
[-\fBs \fIstrength\fR[\fI,strength\fR]]
[-\fBt \fItbfile.tb\fR]
.
.SH DESCRIPTION
\fBdobutsu\fR is an engine for the Japanese Shogi variant
\fIDobutsu Shogi\fR with a
.BR gnuchess (6)
like interface.
.
It uses a comprehensive endgame tablebase to achieve perfect play but
can also play at any desired strength.
.LP
The following options are supported:
.TP
-\fBc\fR \fIcolor\fR
Make the engine play \fIcolor\fR.
.
If \fIcolor\fR is one of \fBbBsS\fR, the engine plays Sente.
.
If it is one of \fBwWgG\fR, the engine plays Gote.
.
More than one colour can be provided to have the engine play against
itself.
.TP
-\fBq\fR
Do not print the board after each move.
.
This is the default.
.TP
-\fBs \fIstrength\fR[\fI,strength\fR]
Set engine strength to \fIstrength\fR, a positive floating point number.
.
All values over 700 cause the engine to play perfectly.
.
If two strength values of the form \fIstrength,strength\fR are provided,
the first strength is used for Sente while the second strength is used
for Gote.
.TP
-\fBt\fR \fItbfile.tb\fR
Load the endgame tablebase from \fItbfile.tb.\fR
.
If no -\fBt\fR option is provided, the endgame tablebase is loaded from
the location indicated in the environment variable DOBUTSU_TABLEBASE.
.
If that variable is unset, files \fIdobutsu.tb\fR and
\fIdobutsu.tb.xz\fR are tried in the current working directory.
.TP
-\fBv\fR
Print the board after each move.
.
.SH COMMANDS
\fBdobutsu\fR is an interactive program that asks the user for commands
and responds to them if requested.
.
All interaction, including error messages, happens on standard input and
standard output.
.
Only fatal error messages are printed to standard error.
.
A command is either a move in algebraic notation or one of the
following:
.TP
\fBboth\fR
Make engine play both players, immediately playing out the rest of the
game at current engine strengths.
.TP
\fBexit\fR
Exit the program.
.TP
\fBforce\fR
Set the engine to play neither Sente nor Gote.
.TP
\fBgo\fR
Make the engine play the colour that is currently on the move.
.TP
\fBhelp\fR
Print a list of commands.
.TP
\fBhint\fR
Print what the engine would play.
.TP
\fBnew\fR [\fIposition\fR]
Start a new game.
.
If \fIposition\fR is provided, use the specified position as the
starting position, otherwise use the default position.  The format of
the position string is adopted from Forsyth-Edwards-Notation.  Refer
to the output of \fBshow position\fR for how this notation works.
.TP
\fBremove\fR
Undo the last two moves.
.TP
\fBsetup\fR [\fIposition\fR]
Alias for \fBnew\fR.
.TP
\fBshow \fIthing\fR
Print various useful pieces of information.
.
Valid values for
\fIthing\fR are:
.RS
.TP
\fBboard\fR
Print the current board.
.TP
\fBsetup\fR
Print board as a position string
.TP
\fBmoves\fR
Print possible moves.
.TP
\fBeval\fR
Print position evaluation.
.TP
\fBlines\fR
Print possible moves and their evaluation.
.RE
.TP
\fBstrength [\fIstrength\fR [\fIstrength\fR]]
If invoked with no arguments, print the current engine strengths.
.IP
If invoked with one or two arguments, set engine strength to a positive
floating-point number.
.
If two strengths are provided, the first one is used for Sente while the
second one is used for Gote.
.
A strength above 700 makes the engine play perfectly.
.TP
\fBundo\fR
Undo the previous move.
.
If the last move was an engine move, the engine is going to play another
move immediately.
.
Either use \fBforce\fR to stop the engine or \fBremove\fR to undo two
moves at a time.
.TP
\fBquiet\fR
Do not print the board after every move.
.TP
\fBquit\fR
Quit the program.
.TP
\fBverbose\fR
Print the board after every move.
.TP
\fBversion\fR
Print program version.
.
.SH RULES
Dobutsu Shogi is a two-player strategy game played on a 3\(mu4
board.
.
Each player has four pieces: A \fIlion,\fR an \fIelephant,\fR a
\fIgiraffe,\fR and a \fIchick\fR.
.
The player whose lion starts on the fourth rank is called \fISente.\fR
.
His pieces are denoted by uppercase characters.
.
The other player is called \fIGote\fR, denoted by lowercase characters.
.
\fISente\fR makes the first move, then both players alternate moves.
.
The initial setup looks like this:\FC
.LP
.RS
.nf
  ABC 
 +---+
1|gle|
2| c |
3| C |
4|ELG|
 +---+
.fi
.RE
.LP
\FTThe goal of the game is to either capture the opponents lion, or to
advance one's own lion to an unattacked square in the \fIpromotion
zone,\fR which is the rank farthest from the lion's starting square.
.
Sente's promotion zone\ is rank 1, whereas Gote's promotion zone is
rank 4.
.LP
Each piece can move in a variety of ways: The lion can move one
square into every direction, the elephant can move one square
diagonally, the giraffe can move one square horizontally or
vertically, the chick can move one square towards the promotion zone.
.
When a chick is moved into the promotion zone, it is promoted to a
\fIrooster,\fR which can move horizontally, vertically, or diagonally
towards the promotion zone.
.
The following diagram shows with asterisks the squares each of
Sente's pieces can move to:\FC
.LP
.RS
.nf
***  * *   *    *   ***
*L*   E   *G*   C   *R*
***  * *   *         *
.fi
.RE
.LP
\FTPieces cannot move onto squares occupied by a player's own pieces.
.
When a piece is moved onto a square occupied by an opponent's piece,
that piece is \fIcaptured\fR and transferred into the \fIhand\fR of the
capturer.
.
If the captured piece is a rooster, it is unpromoted to a chick.
.
Instead of moving a piece, a player may also \fIdrop\fR one of the piece
in his hand onto an arbitrary square on the board.
.LP
If a board position is repeated for the third time, the game is a draw.
.
There is no stalemate; if the only possible move is to move into check,
the lion has to move into check.
.
If no move is possible, the side to move looses the game.
.
.SH ENVIRONMENT
.TP
DOBUTSU_TABLEBASE
The location of the endgame tablebase if no -\fBt \fItbfile.tb\fR option
is specified.  If unset, this defaults to the where the tablebase file
was placed on installation.
.
.SH FILES
.TP
\fIdobutsu.tb, dobutsu.tb.xz\fR
Endgame tablebase files.
.
If no -\fBt \fItbfile.tb\fR option is specified and DOBUTSU_TABLEBASE is
unset, \fBdobutsu\fR tries to open files named \fIdobutsu.tb\fR and
\fIdobutsu.tb.xz\fR in the current working directory.
.
.SH EXIT STATUS
.TP
1
A fatal error occured.
.TP
0
No fatal error occured.
.
.SH EXAMPLES
The following is a sample session with some commands demonstrated.
.
The program prints the current move number as a prompt, user input is
printed in bold.
.LP
.RS
.nf
\FCLoading tablebase... done
1. \fBshow board\fR
  ABC 
 +---+
1|gle| 
2| c |
3| C |
4|ELG| *
 +---+
1. \fBCb3b2\fR
2. \fBshow board\fR
  ABC 
 +---+
1|gle| *
2| C |
3|   |
4|ELG| C
 +---+
2. \fBshow setup\fR
G/gle/-C-/---/ELG/C
2. \fBshow lines\fR
Ec1xb2 : #75   (24.22%)
Lb1xb2 : #-72  (22.95%)
Lb1-c2 : #-30  (22.07%)
Lb1-a2 : #-30  (22.07%)
Ga1-a2 : #-2   ( 8.68%)
2. \fBstrength 20\fR
2. \fBshow lines\fR
Ec1xb2 : #75   (60.59%)
Lb1xb2 : #-72  (20.54%)
Lb1-c2 : #-30  ( 9.44%)
Lb1-a2 : #-30  ( 9.44%)
Ga1-a2 : #-2   ( 0.00%)
2. \fBshow moves\fR
Ga1-a2
Ec1xb2
Lb1-c2
Lb1xb2
Lb1-a2
2. \fBshow eval\fR
#75
2. \fBgo\fR
My 2. move is : Ec1xb2
3. \fBgo\fR
My 3. move is : Gc4-c3
4. \fBshow setup\fR
G/gl-/-e-/-C-/ELG/c
4. \fBquit\fR\FT
.fi
.RE
.
.SH DIAGNOSTICS
.TP
\fBLoading tablebase... \fItbfile.tb: some error\fR
The tablebase could not be loaded for some reason.
.
All functionality that accesses the tablebase is unavailable.
.TP
\fBError (tablebase unavailable) : \fIsome command\fR
The tablebase could not be loaded and you tried to execute a command
that requires the tablebase to be loaded.
.TP
\fBError (invalid position) : \fIsome command\fR
You entered a \fBnew\fR or \fBsetup\fR command with an invalid position
string.
.TP
\fBError (unknown command) : \fIsome command\fR
You entered a command that is not known to this program.
.
Type \fBhelp\fR for a list of commands.
.TP
\fBError (strength must be positive) : \fIsome command\fR
You tried to set the engine strength to a nonpositive value.
.TP
\fBError (invalid strength) : \fIsome command\fR
You used the \fBstrength\fR command with an argument that could not be
parsed into a floating point number.
.
.SH SEE ALSO
.BR gnuchess (6),
.BR gnugo (6),
.BR xboard (6)
.
.SH CAVEATS
Even though its user interface is very similar to
.BR gnuchess (6),
.BR dobutsu (6)
is not
.BR xboard (6)
compatible.
.LP
Stricly speaking, the Dobutsu Shogi rules allow either play to
move his Lion to an attacked square on the fourth rank without winning.
.
The lion can then immediately be captured by the opponent.
.
As the rules are not clear as to what happens when the opponent does not
immediately capture, this implementation does not allow the lion to move
onto an attacked square on the fourth rank.
.
Moving into check is allowed in other situations.
.LP
The -\fBs \fIstrength\fR option adjusts a weighting function used during
move selection.
.
As the only information the engine has about the position's evaluation
is the entry in the endgame tablebase, its notion of what is a good move
sometimes differs from what a human thinks.
.
.SH HISTORY
Dobutsu Shogi is a Shogi variant invented by professional
Shogi player Madoka Kitao in 2008 and solved by University of Tokyo
researcher Tanaka Tetsuro in 2009.
.
Development of this software began as a hobby project in
July 2016 and concluded in February 2017.
.
.SH AUTHOR
.nf
Robert Clausecker <fuz@fuz.su>
.fi


CHANGES
2021-01-16 dobutsu-2

 * new command "show setup"
 * readline/libedit support
 * fix command "strength" and make option -c work
 * locale support
 * improved build scripts
 * improved documentation

2017-02-12 dobutsu-1

 * initial release


dobutsu.c
/*-
 * Copyright (c) 2016--2017, 2021 Robert Clausecker. All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */
#define _POSIX_C_SOURCE 200809L
#include <assert.h>
#include <errno.h>
#include <locale.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include <readline/readline.h>
#include <readline/history.h>

#include "rules.h"
#include "tablebase.h"
#include "version.h"

/*
 * The struct game represents the entire state of the game.
 */
struct gamestate {
	struct gamestate *previous;
	struct position position;
	struct move next_move;
	unsigned move_clock;
};

/*
 * The engine may play any combination of Sente and Gote, including
 * no player.
 */
enum {
	ENGINE_NONE = 0,
	ENGINE_SENTE = 1,
	ENGINE_GOTE = 2,
	ENGINE_BOTH = 3
};

/* global variables */
static struct tablebase *tb = NULL;
static struct gamestate *gs = NULL;
static unsigned char engine_players = 0;
static unsigned char show_board_after_move = 0;
static double sente_strength = 1, gote_strength = 1;
static struct seed seed;
static char *linebuf = NULL;

/* internal functions */
static void	open_tablebase(const char *);
static void	execute_command(char *);
static void	end_game(void);
static void	cmd_hint(const char *);
static void	cmd_new(const char *);
static void	cmd_exit(const char *);
static void	cmd_show(const char *);
static void	cmd_show_board(void);
static void	cmd_show_moves(void);
static void	cmd_show_eval(void);
static void	cmd_show_lines(void);
static void	cmd_show_setup(void);
static void	cmd_strength(const char *);
static void	cmd_undo(const char *);
static void	cmd_remove(const char *);
static void	cmd_help(const char *);
static void	cmd_version(const char *);
static void	cmd_both(const char *);
static void	cmd_go(const char *);
static void	cmd_force(const char *);
static void	cmd_verbose(const char *);
static void	cmd_quiet(const char *);
static int	play(struct move m);
static char	*prompt(void);
static void	autoplay(void);
static int	undo(void);
static int	draw(void);
static void	error(const char *);

/*
 * This table contains all available commands.  New commands should be
 * added to this table.  The callback function takes as its sole
 * argument a pointer to the remainder of the command line.
 * The table is terminated with an entry containing NULL for the
 * function pointer.
 */
static const struct {
	void (*callback)(const char *);
	char command[8];
} commands[] = {
	cmd_both,	"both",
	cmd_exit,	"exit",
	cmd_force,	"force",
	cmd_go,		"go",
	cmd_help,	"help",
	cmd_hint,	"hint",
	cmd_new,	"new",
	cmd_exit,	"quit",
	cmd_quiet,	"quiet",
	cmd_remove,	"remove",
	cmd_new,	"setup",
	cmd_new,	"setboard",
	cmd_show,	"show",
	cmd_strength,	"strength",
	cmd_undo,	"undo",
	cmd_verbose,	"verbose",
	cmd_version,	"version",
	NULL,		""
};

/*
 * Similar to the previous table, this table contains all subcommands
 * for the show command.
 */
static const struct {
	void (*callback)(void);
	char command[8];
} show_commands[] = {
	cmd_show_board,	"board",
	cmd_show_moves, "moves",
	cmd_show_eval,	"eval",
	cmd_show_lines,	"lines",
	cmd_show_setup,	"setup",
	NULL,		""
};

extern int
main(int argc, char *argv[])
{
	int optchar;
	unsigned char players = 0;
	char *tbloc = getenv("DOBUTSU_TABLEBASE");

	setlocale(LC_ALL, "");

	while (optchar = getopt(argc, argv, "c:qs:t:v"), optchar != EOF)
		switch (optchar) {
		case 'c':
			while (*optarg != '\0')
				switch (*optarg++) {
				case 'b':
				case 'B':
				case 's':
				case 'S':
					players |= ENGINE_SENTE;
					break;

				case 'w':
				case 'W':
				case 'g':
				case 'G':
					players |= ENGINE_GOTE;
					break;

				default:
					fprintf(stderr, "Cannot play for %c\n", *optarg);
					return (EXIT_FAILURE);
				}

			break;

		case 'q':
			show_board_after_move = 0;
			break;

		case 'v':
			show_board_after_move = 1;
			break;

		case 's':
			switch (sscanf(optarg, "%lf,%lf", &sente_strength, &gote_strength)) {
			case 1:
				gote_strength = sente_strength;
				break;

			case 2:
				break;

			default:
				fprintf(stderr, "Cannot parse strength: %s\n", optarg);
				return (EXIT_FAILURE);
			}

			if (!(gote_strength > 0 && sente_strength > 0)) {
				fprintf(stderr, "Strength must be positive: %s\n", optarg);
				return (EXIT_FAILURE);
			}

			break;

		case 't':
			tbloc = optarg;
			break;

		case ':':
		case '?':
		default:
			return (EXIT_FAILURE);
		}

	/* for better interaction */
	setbuf(stdin, NULL);
	setbuf(stdout, NULL);

	ai_seed(&seed);
	open_tablebase(tbloc);
	cmd_new("");

	engine_players = players;
	autoplay();

	using_history();

	while (linebuf = readline(prompt()), linebuf != NULL) {
		execute_command(linebuf);
		add_history(linebuf);
		free(linebuf);
		linebuf = NULL;
	}

	cmd_exit(NULL);
}

/*
 * Open the endgame tablebase in file tbloc.  If tbloc is NULL,
 * try opening a file named dobutsu.tb in the current working
 * directory.  If that doesn't work either, give up.
 */
static void
open_tablebase(const char *tbloc)
{
	FILE *tbfile;

	printf("Loading tablebase... ");

	if (tbloc != NULL)
		tbfile = fopen(tbloc, "rb");
	else {
		tbloc = "dobutsu.tb";
		tbfile = fopen(tbloc, "rb");
		if (tbfile == NULL && errno == ENOENT) {
			tbloc = "dobutsu.tb.xz";
			tbfile = fopen(tbloc, "rb");
		}
	}

	if (tbfile != NULL) {
		tb = read_tablebase(tbfile);
		fclose(tbfile);
	}

	if (tb != NULL)
		puts("done");
	else
		printf("%s: %s\n", tbloc, errno == 0 ? "Unknown error" : strerror(errno));
}

/*
 * The error function prints a string of the form
 * "Error (msg): command" to stdout to signalize that a command failed.
 * command is taken from linebuf.
 */
static void
error(const char *msg)
{

	printf("Error (%s) : %s\n", msg, linebuf == NULL ? "" : linebuf);
}

/*
 * Deallocate gs and the entire undo-state corresponding to it.
 */
static void
end_game()
{
	struct gamestate *gsptr = gs, *prevgs;

	while (gsptr != NULL) {
		prevgs = gsptr->previous;
		free(gsptr);
		gsptr = prevgs;
	}
}

/*
 * Start a new game by clearing the old game state and initializing it
 * with the state of a new game.  If the first argument is empty, use
 * the default setup.  If the first argument is not empty, try to setup
 * the board with the specified setup.
 */
static void
cmd_new(const char *arg)
{
	struct position p;

	if (arg[0] == '\0')
		p = INITIAL_POSITION;
	else if (parse_position(&p, arg) != 0) {
		error("invalid position");
		return;
	}

	end_game();
	engine_players = ENGINE_NONE;
	gs = malloc(sizeof *gs);
	if (gs == NULL) {
		perror("malloc");
		exit(EXIT_FAILURE);
	}

	gs->previous = NULL;
	gs->position = p;
	gs->move_clock = 1;
}

/*
 * Execute the command given in cmd.
 */
static void
execute_command(char *cmd)
{
	struct move m;
	size_t i, cmdlen;
	char *arg;

	/* trim newline if any */
	cmd[strcspn(cmd, "\r\n")] = '\0';


	/* trim leading whitespace */
	cmd += strspn(cmd, " ");

	/* if a move is given, try to play that move */
	if (parse_move(&m, &gs->position, cmd) == 0) {
		if (play(m)) {
			puts("You win!\nStarting new game.");
			cmd_new("");
		} else if (draw()) {
			puts("Draw by threefold repetition.\nStarting new game.");
			cmd_new("");
		}

		autoplay();
		return;
	}

	/* else, split command at the first whitespace */
	cmdlen = strcspn(cmd, " ");
	arg = cmd + cmdlen + strspn(cmd + cmdlen, " ");

	/* do nothing on empty command */
	if (cmdlen == 0)
		return;

	if (cmdlen <= sizeof commands[i].command)
		for (i = 0; commands[i].callback != NULL; i++) {
			if (strncmp(commands[i].command, cmd, cmdlen) != 0)
				continue;

			if (strnlen(commands[i].command, sizeof commands[i].command) != cmdlen)
				continue;

			commands[i].callback(arg);
			return;
		}

	error("unknown command");
}

/*
 * Play m on the current game state and update gs appropriately.
 * Return nonzero if the game ended through that move.
 */
static int
play(struct move m)
{
	struct gamestate *newgs = malloc(sizeof *newgs);
	int game_ends;

	if (newgs == NULL) {
		perror("malloc");
		exit(EXIT_FAILURE);
	}

	newgs->previous = gs;
	newgs->position = gs->position;
	newgs->move_clock = gs->move_clock + 1;
	gs->next_move = m;
	gs = newgs;

	game_ends = play_move(&gs->position, &m);

	if (show_board_after_move)
		cmd_show_board();

	return (game_ends);
}

/*
 * Return nonzero if it's the engine's turn.
 */
static int
engine_moves(void)
{
	return engine_players &
	    (gote_moves(&gs->position) ? ENGINE_GOTE : ENGINE_SENTE);
}

/*
 * If it's the engine's turn, play a move for the engine.  Repeat until
 * either it's no longer the engine's turn or (if both players are
 * engine players) until the game ends.
 */
static void
autoplay(void)
{
	struct move engine_move;
	double strength;
	int end;
	unsigned old_clock;
	char movstr[MAX_MOVSTR];

	while (engine_moves()) {
		if (tb == NULL) {
			error("tablebase unavailable");
			engine_players = ENGINE_NONE;
			return;
		}

		strength = gote_moves(&gs->position) ? gote_strength : sente_strength;

		engine_move = ai_move(tb, &gs->position, &seed, strength);
		move_string(movstr, &gs->position, &engine_move);
		old_clock = gs->move_clock;
		end = play(engine_move);
		printf("My %u. move is : %s\n", old_clock, movstr);
		if (end) {
			puts("I win!\nStarting new game.");
			cmd_new("");
		} else if (draw()) {
			puts("Draw by threefold repetition.\nStarting new game.");
			cmd_new("");
		}
	}
}

/*
 * Generate an engine move for the current colour and print it out.
 */
static void
cmd_hint(const char *arg)
{
	struct move aim;
	double strength = gote_moves(&gs->position) ? gote_strength : sente_strength;
	char movstr[MAX_MOVSTR];

	(void)arg;

	if (tb == NULL) {
		error("tablebase unavailable");
		return;
	}

	aim = ai_move(tb, &gs->position, &seed, strength);
	move_string(movstr, &gs->position, &aim);
	puts(movstr);
}

/*
 * Close everything and exit the program.  This function should not
 * return.  The first argument is ignored.
 */
static void
cmd_exit(const char *arg)
{

	(void)arg;

	end_game();
	free_tablebase(tb);
	tb = NULL;
	free(linebuf);

	if (ferror(stdin)) {
		perror("Error reading command");
		exit(EXIT_FAILURE);
	}

	exit(EXIT_SUCCESS);
}

/*
 * The show command has several subcommands.  Due to the small number of
 * subcommands, we call them directly through an if-cascade.
 */
static void
cmd_show(const char *arg)
{
	size_t i;

	for (i = 0; show_commands[i].callback != NULL; i++) {
		if (strcmp(arg, show_commands[i].command) != 0)
			continue;

		show_commands[i].callback();
		return;
	}

	error("unknown command");
}

/*
 * Print the current board to stdout.
 */
static void
cmd_show_board()
{
	char render[MAX_RENDER];

	position_render(render, &gs->position);
	fputs(render, stdout);
}

/*
 * Print all possible moves to stdout.
 */
static void
cmd_show_moves()
{
	struct move moves[MAX_MOVES];
	size_t i, n;
	char movstr[MAX_MOVSTR];

	n = generate_moves(moves, &gs->position);
	for (i = 0; i < n; i++) {
		move_string(movstr, &gs->position, moves + i);
		puts(movstr);
	}
}

/*
 * Lookup the current position in the tablebase and print its
 * evaluation.
 */
static void
cmd_show_eval(void)
{
	tb_entry eval;

	if (tb == NULL) {
		error("tablebase unavailable");
		return;
	}

	eval = lookup_position(tb, &gs->position);
	if (is_win(eval))
		printf("#%d\n", get_dtm(eval));
	else if (is_loss(eval))
		printf("#-%d\n", get_dtm(eval));
	else /* is_draw(eval) */
		printf("0\n");
}

/*
 * For each move, print its evaluation and engine probability.
 */
static void
cmd_show_lines(void)
{
	struct analysis analysis[MAX_MOVES];
	double strength = gote_moves(&gs->position) ? gote_strength : sente_strength;
	size_t i, nmove;
	char movstr[MAX_MOVSTR], dtmstr[6];

	if (tb == NULL) {
		error("tablebase unavailable");
		return;
	}

	nmove = analyze_position(analysis, tb, &gs->position, strength);
	for (i = 0; i < nmove; i++) {
		move_string(movstr, &gs->position, &analysis[i].move);
		if (is_draw(analysis[i].entry))
			strcpy(dtmstr, "0");
		else
			snprintf(dtmstr, sizeof dtmstr,
			    is_win(analysis[i].entry) ? "#%d" : "#-%d",
			    get_dtm(analysis[i].entry));

		printf("%-7s: %-5s (%5.2f%%)\n", movstr, dtmstr, analysis[i].value * 100.0);
	}
}

/*
 * Print board configuration as position string.
 */
static void
cmd_show_setup(void)
{
	char render[MAX_POSSTR];

	position_string(render, &gs->position);
	puts(render);
}

/*
 * The strength command lets you set the engine strength.  If no operand
 * is provided, the current engine strength is printed.  If one operand
 * is provided, both engine's strengths are set to that value.  If two
 * operands are provided, Sente's and Gote's strengths are set to the
 * two values.
 */
static void
cmd_strength(const char *arg)
{
	double s, g;

	switch (sscanf(arg, "%lf%lf", &s, &g)) {
	case 1:
		g = s;
		/* fallthrough */

	case 2:
		/* also catch NaN */
		if (!(s > 0 && g > 0)) {
			error("strength must be positive");
			return;
		}

		sente_strength = s;
		gote_strength = g;
		break;

	/* there was an argument but no %lf could be parsed */
	case 0:
		error("invalid strength");
		break;

	/* there was no argument */
	case EOF:
	default:
		printf("Sente: %6.2f\nGote:  %6.2f\n", sente_strength, gote_strength);
		break;

	}
}

/*
 * The undo command lets the player undo a single move.
 */
static void
cmd_undo(const char *arg)
{

	(void)arg;

	if (undo())
		autoplay();
}

/*
 * The remove command lets the player undo two moves.  This is
 * useful when you are playing against the engine.
 */
static void
cmd_remove(const char *arg)
{

	(void)arg;

	if (undo() && undo())
		autoplay();
}

/*
 * undo undoes the last move played if such a move existed.  This
 * function returns nonzero if a move to be undone exists.
 */
static int
undo(void)
{
	struct gamestate *oldgs = gs;

	if (oldgs->previous != NULL) {
		gs = oldgs->previous;
		free(oldgs);
		return (1);
	} else {
		printf("Nothing to undo.\n");
		return (0);
	}
}


/*
 * Print a list of commands. Argument is ignored.
 */
static void
cmd_help(const char *arg)
{

	(void)arg;

	printf(
	    "help        print a list of commands\n"
	    "hint        print what the engine would play\n"
	    "exit        leave the program\n"
	    "version     print program version\n"
	    "new         start a new game\n"
	    "undo        undo previous move\n"
	    "remove      undo last two moves\n"
	    "setup       setup board with position string\n"
	    "show board  print the current board\n"
	    "show setup  print board as a position string\n"
	    "show moves  print possible moves\n"
	    "show eval   print position evaluation\n"
	    "show lines  print possible moves and their evaluations\n"
	    "strength    show/set engine strength\n"
	    "both        make engine play both players\n"
	    "go          make the engine play the colour that is on the move\n"
	    "force       set the engine to play neither colour\n"
	    "verbose     print the board after every move\n"
	    "quiet       do not print the board after every move\n");
}

/*
 * Print the program version.  The version can be set using the VERSION
 * macro.  If VERSION is unset, "unknown" is assumed. The argument is
 * ignored.  Argument is ignored.
 */
static void
cmd_version(const char *arg)
{

	(void)arg;
	printf("dobutsu " DOBUTSU_VERSION "\n");
	printf("Copyright (c) 2016--2017, 2021 Robert Clausecker <fuz@fuz.su>\nAll rights reserved.\n");
}

/*
 * Make the program play whatever players move it currently is and make
 * the human take the other colour.  Then play a move.  The argument is
 * ignored.
 */
static void
cmd_go(const char *arg)
{

	(void)arg;

	engine_players = gote_moves(&gs->position) ? ENGINE_GOTE : ENGINE_SENTE;
	autoplay();
}

/*
 * Set both engines to run, arg is ignored.
 */
static void
cmd_both(const char *arg)
{

	(void)arg;

	engine_players = ENGINE_BOTH;
	autoplay();
}

/*
 * Set both players to manual.  The argument is ignored.
 */
static void
cmd_force(const char *arg)
{

	(void)arg;

	engine_players = ENGINE_NONE;
}

/*
 * Generate a prompt of the form "##. " where ## is the current move number.
 */
static char *
prompt(void)
{
	static char promptstr[13];

	snprintf(promptstr, sizeof promptstr, "%u. ", gs->move_clock);
	return (promptstr);
}

/*
 * Check if the current position is a draw by threefold repetition.
 * Return nonzero if it is, zero otherwise.
 */
static int
draw(void)
{
	struct position *p = &gs->position;
	struct gamestate *gsptr;
	int repetitions = 1;

	for (gsptr = gs->previous; gsptr != NULL; gsptr = gsptr->previous)
		if (position_equal(p, &gsptr->position) && ++repetitions == 3)
			return (1);


	return (0);
}

/*
 * Turn on printing the board after every move.
 */
static void
cmd_verbose(const char *arg)
{

	(void)arg;
	show_board_after_move = 1;
}

/*
 * Turn off printing the board after every move.
 */
static void
cmd_quiet(const char *arg)
{

	(void)arg;
	show_board_after_move = 0;
}
